thrift/src/main/java/com/linecorp/armeria/server/thrift/ThriftServiceSpecificationGenerator.java

```diff
@@ -0,0 +1,465 @@
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nullable;
+
+import org.apache.thrift.TBase;
+import org.apache.thrift.TException;
+import org.apache.thrift.TFieldIdEnum;
+import org.apache.thrift.TFieldRequirementType;
+import org.apache.thrift.TSerializer;
+import org.apache.thrift.meta_data.EnumMetaData;
+import org.apache.thrift.meta_data.FieldMetaData;
+import org.apache.thrift.meta_data.FieldValueMetaData;
+import org.apache.thrift.meta_data.ListMetaData;
+import org.apache.thrift.meta_data.MapMetaData;
+import org.apache.thrift.meta_data.SetMetaData;
+import org.apache.thrift.meta_data.StructMetaData;
+import org.apache.thrift.protocol.TType;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.annotations.VisibleForTesting;
+
+import com.linecorp.armeria.common.thrift.ThriftProtocolFactories;
+import com.linecorp.armeria.server.ServiceConfig;
+import com.linecorp.armeria.server.docs.ClassInfo;
+import com.linecorp.armeria.server.docs.CollectionInfo;
+import com.linecorp.armeria.server.docs.EndpointInfo;
+import com.linecorp.armeria.server.docs.EnumInfo;
+import com.linecorp.armeria.server.docs.ExceptionInfo;
+import com.linecorp.armeria.server.docs.FieldInfo;
+import com.linecorp.armeria.server.docs.FieldRequirement;
+import com.linecorp.armeria.server.docs.FunctionInfo;
+import com.linecorp.armeria.server.docs.ListInfo;
+import com.linecorp.armeria.server.docs.MapInfo;
+import com.linecorp.armeria.server.docs.ServiceInfo;
+import com.linecorp.armeria.server.docs.ServiceSpecification;
+import com.linecorp.armeria.server.docs.ServiceSpecificationGenerator;
+import com.linecorp.armeria.server.docs.SetInfo;
+import com.linecorp.armeria.server.docs.StructInfo;
+import com.linecorp.armeria.server.docs.TypeInfo;
+
+/**
+ * {@link ServiceSpecificationGenerator} implementation that supports {@link THttpService}s.
+ */
+public class ThriftServiceSpecificationGenerator implements ServiceSpecificationGenerator {
+
+    private static final ObjectMapper mapper = new ObjectMapper();
+
+    @Override
+    public ServiceSpecification generate(Iterable<ServiceConfig> serviceConfigs) {
+        final Map<Class<?>, Iterable<EndpointInfo>> map = new LinkedHashMap<>();
+
+        for (ServiceConfig c : serviceConfigs) {
+            final Optional<THttpService> serviceOpt = c.service().as(THttpService.class);
+            if (!serviceOpt.isPresent()) {
+                continue;
+            }
+
+            final THttpService service = serviceOpt.get();
```

Comment:
Sounds like a good idea. Let me update this (as well as addressing your other comments)